[WIP] JIT Static Hooks: serialization logic

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49547 [WIP] JIT Static Hooks: cpp tests
* #49546 [WIP] JIT Static Hooks: python tests
* **#49545 [WIP] JIT Static Hooks: serialization logic**
* #49975 [WIP] JIT Static Hooks: schema checking logic
* #49544 [WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation

Adds logic to enable serialization/deserialization of modules with forward hooks and pre-hooks. 

Hooks are printed out like a module's method inside of the module's serialized code body.

Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

Commit:
[WIP] JIT Static Hooks: serialization logic

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/serialization/import_source.cpp:
```diff
@@ -367,13 +367,22 @@ struct SourceImporterImpl : public Resolver,
     std::vector<Def> methods;
-    std::vector<ResolverPtr> resolvers;
+    std::vector<ResolverPtr> method_resolvers;
+    std::map<std::string, Def> pre_hook_def_map;
+    std::map<std::string, Def> hook_def_map;
+    std::map<std::string, ResolverPtr> pre_hook_resolver_map;
+    std::map<std::string, ResolverPtr> hook_resolver_map;
```
Do these maps need to be ordered?

Lilyjjo replied:
no because the ordering is gotten from the hook_names/pre_hook_names vectors

SplitInfinity commented on torch/csrc/jit/serialization/import_source.cpp:
```diff
@@ -408,6 +417,25 @@ struct SourceImporterImpl : public Resolver,
                 for (const auto& buffer : buffer_list) {
                   buffer_names.insert(StringLiteral(buffer).text());
                 }
+              } else if (name == "__forward_pre_hooks__") {
+                TORCH_INTERNAL_ASSERT(
+                    is_module,
+                    "Forward pre hooks only exist on modules at the moment");
+                const auto pre_hook_list =
+                    ListLiteral(assign.rhs().get()).inputs();
+                for (const auto& pre_hook : pre_hook_list) {
+                  pre_hook_names.insert(StringLiteral(pre_hook).text());
+                  pre_hooks_order.emplace_back(StringLiteral(pre_hook).text());
+                }
+              } else if (name == "__forward_hooks__") {
+                TORCH_INTERNAL_ASSERT(
+                    is_module,
+                    "Forward hooks only exist on modules at the moment");
+                const auto hook_list = ListLiteral(assign.rhs().get()).inputs();
+                for (const auto& hook : hook_list) {
+                  hook_names.insert(StringLiteral(hook).text());
+                  hooks_order.emplace_back(StringLiteral(hook).text());
+                }
```
This will likely pose forward compatibility problems. I think we need to gate this with `#ifndef FBCODE_CAFFE2` and then remove the guard in a few months time.

cc: @gmagogsfm, exact same problem that caused the SEV we discussed.

Lilyjjo replied:
What issue is this gate trying to solve/prevent? The TK_DEF case below on 473 will treat the pre_hooks as methods if they're not added to the list here. I might have to gate things in python_print.cpp as well

Lilyjjo replied:
I added the \#ifndef to the part where the hooks are printed into the IR, so this code won't be triggered. Would that take care of the forward compatibility issue?

SplitInfinity replied:
Yes - and you're right, we don't need to add the macro here, only n `python_print.cpp`.

SplitInfinity commented on torch/csrc/jit/serialization/import_source.cpp:
```diff
@@ -367,13 +367,22 @@ struct SourceImporterImpl : public Resolver,
     std::unordered_set<std::string> buffer_names;
+    std::unordered_set<std::string> pre_hook_names;
+    std::unordered_set<std::string> hook_names;
+    // use to keep track of original ordering of hooks and prehooks
```
```suggestion
    // use to keep track of original ordering of hooks and prehooks in case any are called more than once
```

SplitInfinity commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1394,6 +1412,19 @@ struct PythonPrintImpl {
       for (auto& method : classType->methods()) {
         printFunction(*method);
       }
+      std::set<std::string> already_printed;
+      for (auto& hook : classType->getForwardHooks()) {
+        if (already_printed.count(hook->name()) == 0) {
+          already_printed.insert(hook->name());
+          printFunction(*hook);
+        }
+      }
+      for (auto& pre_hook : classType->getForwardPreHooks()) {
+        if (already_printed.count(pre_hook->name()) == 0) {
+          already_printed.insert(pre_hook->name());
+          printFunction(*pre_hook);
+        }
+      }
```
It's probably a corner case but this might lead to bloating of serialized file size if the same hook is registered on multiple modules that are saved into the same archive.

Lilyjjo replied:
Is the bloat that much of a concern? The hooks should take up the same amount of space as other methods on a module

SplitInfinity replied:
Sure, but won't they be repeated in the body of the definition of every module that uses them?

I guess we can emit them inside the module definition for now, but let's keep this in mind in case the file size of a serialized archive of a module with hooks attached becomes problematic in the future. We should probably file an issue so that we don't lose track of it.

SplitInfinity reviewed (commented):
There is forward compatibility issue in this PR that we have to be very careful about. We can chat more via direct message.

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Lilyjjo commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1392,6 +1413,19 @@ struct PythonPrintImpl {
       for (auto& method : classType->methods()) {
         printFunction(*method);
       }
+      std::set<std::string> already_printed;
```
Oh, I need to add the /#ifndef here too, will add

SplitInfinity commented on torch/csrc/jit/serialization/import_source.cpp:
```diff
@@ -408,6 +417,25 @@ struct SourceImporterImpl : public Resolver,
+                const auto pre_hook_list =
+                    ListLiteral(assign.rhs().get()).inputs();
+                for (const auto& pre_hook : pre_hook_list) {
+                  pre_hook_names.insert(StringLiteral(pre_hook).text());
```
Consider refactoring `StringLiteral(pre_hook).text()` to a separate variable to avoid repeating it.

SplitInfinity commented on torch/csrc/jit/serialization/import_source.cpp:
```diff
@@ -408,6 +417,25 @@ struct SourceImporterImpl : public Resolver,
+                const auto hook_list = ListLiteral(assign.rhs().get()).inputs();
+                for (const auto& hook : hook_list) {
+                  hook_names.insert(StringLiteral(hook).text());
+                  hooks_order.emplace_back(StringLiteral(hook).text());
```
Same here.

SplitInfinity commented on torch/csrc/jit/frontend/sugared_value.cpp:
```diff
@@ -625,6 +625,13 @@ std::shared_ptr<SugaredValue> ClassValue::attr(
+  // Allow import_source.cpp to resolve calls to a submodule's
+  // hooks. Edge case because normally you wouldn't allow a module to
+  // call functions of a submodule
+  if (Function* hook = type_->findHook(field)) {
+    return std::make_shared<FunctionValue>(hook);
+  }
```
I don't understand the reason this is needed. Under what circumstances are hooks accessed by name?

Lilyjjo replied:
It's because we're calling hooks as functions of submodules instead of methods. From how the serialization is currently implemented, when trying to resolve the function in the graph it see's that the function is stored as an attribute on the submodule's class type and it tried to resolve it. Here's a paste that might help: https://www.internalfb.com/intern/paste/P160808038/

SplitInfinity replied:
> From how the serialization is currently implemented

By this, do you mean that it is a consequence of serializing hooks as if they were methods (and probably using the class name as part of the prefix for the hook function when compiling it)?. We are essentially baking hooks into the type of a module, so this isn't wrong, but I'm planning to add static methods to classes as part of improving TorchBind and that *might* conflict with this. However, I think it's valid to say that hooks are part of the type but not part of the instance (just like static methods), and so if you have a static method named `foo` and you try to attach and script a hook with the same name, compilation will fail (due to changes you made in the first PR in this stack). 

I think this is another thing I'd like to ask for more opinions on. Out of curiosity (👀 ), do you have any idea how much work it would be to change this PR to serialize hooks as separate functions instead of as part of the class definition for a module?

SplitInfinity reviewed (commented):
Nice! I think we should file a follow-up issues for two things:
- moving the hook definitions outside of module definitions in the archive to save space in the case that a hook is used by multiple modules in the same archive (just so we don't forget)
- removing the `#ifdef CAFFE2_FBCODE` after a few months

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1346,6 +1346,27 @@ struct PythonPrintImpl {
+            body_ << "\"" << hook->name() << "\", ";
+          }
+          body_ << "]\n";
+#endif
```
I think this `#endif` needs to be on the line below?

SplitInfinity commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1392,6 +1413,24 @@ struct PythonPrintImpl {
         printFunction(*method);
       }
+#ifndef FBCODE_CAFFE2
+      // TODO Change to print hooks outside of module if same hook
+      // is being used on multiple modules and serialzed file size
+      // is bloating because of it
```
I think you made an issue for this already so I'd prefer not to have a TODO in the code.

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Hooks are printed out like a module's methods are inside of the module's body.

Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Hooks are printed out like a module's methods are inside of the module's body.

Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

SplitInfinity reviewed (approved):
I think this looks good (plus the change to `Module::save` in the PR below). But let's be careful and pay attention to fbcode internal test signals.

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Hooks are printed out like a module's methods are inside of the module's body.

Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Hooks are printed out like a module's methods are inside of the module's body.

Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Hooks are printed out like a module's methods are inside of the module's body.

Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Adds logic to enable serialization/deserialization of modules with forward hooks and pre-hooks. 

Hooks are printed out like a module's method inside of the module's serialized code body.

Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Adds logic to enable serialization/deserialization of modules with forward hooks and pre-hooks. 

Hooks are printed out like a module's method inside of the module's serialized code body.

Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Adds logic to enable serialization/deserialization of modules with forward hooks and pre-hooks. 

Hooks are printed out like a module's method inside of the module's serialized code body.

Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: serialization logic"


Adds logic to enable serialization/deserialization of modules with forward hooks and pre-hooks. 

Hooks are printed out like a module's method inside of the module's serialized code body.

Differential Revision: [D25771121](https://our.internmc.facebook.com/intern/diff/D25771121)

[ghstack-poisoned]
